docs/DesignDocs/Protected_Code_Loader_Proposal.md

```diff
@@ -0,0 +1,161 @@
+    the encrypted enclave image need to be decrypted and than perform the relocation
+    operation.
+
+## Encryption Algorithm in Protected Code Loader
+Using openssl AES-256-GCM as the encryption/decryption algorithm.
+
+## ELF sections Left plaintext
+### PCL table entry 
+A PCL table entry (a section called ".pcltbl" in ELF file)is built into the Enclave 
+image by linking to PCL lib, i.e. liboepcl.a, this part contains enclave decryption
+info on enclave loading and must remain plaintext.
+
+Definition of PCL entry(section ".pcltbl"):
+typedef struct pcl_table_t_
+{
+	/* Current state of PCL: initailized to PCL_PLAIN */
+    pcl_status_e pcl_state;                   
+    uint32_t     reserved1[3];                /* Must be 0 */
+	// GUID must match GUID in Sealed blob
+    uint8_t      pcl_guid[SGX_PCL_GUID_SIZE]; 
+    size_t       sealed_blob_size;            
+    uint32_t     reserved2[2];                /* Must be 0 */
+	/* For security, sealed blob is copied into enclave */
+    uint8_t      sealed_blob[PCL_SEALED_BLOB_SIZE];
+	/* SHA256 digest of decryption key */
+    uint8_t      decryption_key_hash[SGX_SHA256_HASH_SIZE];
+	/* Number of RVAs */
+    uint32_t     num_rvas;                    
+    uint32_t     reserved3[3];                /* Must be 0 */
+	/* Array of rva_size_tag_iv_t */
+    rva_size_tag_iv_t rvas_sizes_tags_ivs[PCL_MAX_NUM_ENCRYPTED_SECTIONS]; 
+}pcl_table_t;
+
+### Sections left plaintext
+1. ELF header  - binary header
+2. Sections table
+3. Segments table
+4. Sections' names string table pointed by e_shstrndx(e.g. .shstrtab)
+5. .oeinfo section holds enclave's metadata(properties)
+6. .bss and .tbss
+7. sections required to construct dyn_info (.dynamic)
+8. sections holds the content provided by entries with index DT_SYMTAB, DT_STRTAB and DT_REL in
+   dyn_info (e.g. .dynsym, .dynstr, .rela.dyn)
+9. sections containing PCL code and data:
+   a. section ".pcltbl"  // Designated section for PCL table
+   b. .nipx, .nipd, .niprod, .nipd_rel, .nipd_rel_ro_local
+10. sections for debugging
+   .comment, .debug_abbrev, .debug_aranges, .debug_info, .debug_line, .debug_lc, .debug_ranges,
+   .debug_str
+
+## Elf sections to be encrypted
+Sections not mentioned above are sections containing IP information that need to be protected.
+Mainly those sections are:
+1. Code sections -- .text in elf file
+2. Data sections -- .data in elf file (initialized local/global variables)
+3. Read-only Data sections -- .rodata in elf file (const variables)
+4. sections containing relocation info related with the above items
+
+## Sealing/Unsealing the decryption key
+To deliver the decryption key in a secure way, the key used to encrypt the enclave should get sealed
+by key policy OE_SEAL_POLICY_PRODUCT on deliverting to Protected Code Loader for decryption in a 
+sealed_blob. The protected code loader will unseal this sealed_blob to get the decryption key. 
```

Comment:
FYI, in some other TEEs, the loader can not get the decryption key, it can _use_ the key to decrypt but can't _get_ the key.  E.g., the key is in a TPM.

Reply:
It might be helpful to review the PCL design in Intel SGX SDK. https://download.01.org/intel-sgx/sgx-linux/2.11/docs/Intel_SGX_Developer_Reference_Linux_2.11_Open_Source.pdf , page 78-86 talks about PCL support. Intel SGX SDK's PCL sample code for the "utility enclave" seals a hardcoded "mock" key and call out "don't use in release code". https://github.com/intel/linux-sgx/blob/74b24b69f90dee2638ab646d9090cb286ec8abc9/SampleCode/SampleEnclavePCL/Seal/Seal.cpp#L37

Reply:
It would be good to add the references directly into the doc.